### PR TITLE
Make tests less brittle w.r.t. upstream changes to warnings

### PR DESCRIPTION
**Pull request by Balandat**
Mitigates #824.

A proper solution should not count warnings, but check for presence/non-presence of the specific warning, see e.g. change in test_cholesky.py here.

### CODE REVIEW

**Comment by gpleiss** [comment]
@Balandat - I just pushed a commit which tests for specific warnings on all the unit tests in question. If it looks good to you I think this can be merged :)

Closes #824

**Comment by Balandat** [comment]
So the failures on master are caused by #822. 

You can't see that because all the the DeprecationWarnings completely spam the test log. Unfortunately this is a breaking change so we can't just use bool dtype on both stable and master. 

One thing we could do is filter out the specific DeprecationWarning related to this change...

File: gpytorch/variational/whitened_variational_strategy.py [hunk 1]
```diff
@@ -189,8 +189,9 @@ def forward(self, x):
             if self.training:
                 data_covariance = DiagLazyTensor((data_data_covar.diag() - interp_data_data_var).clamp(0, math.inf))
             else:
-                neg_induc_data_data_covar = induc_induc_covar.inv_matmul(
-                    induc_data_covar, left_tensor=induc_data_covar.transpose(-1, -2).mul(-1)
+                neg_induc_data_data_covar = torch.matmul(
```
Comment (Balandat) [inline gpytorch/variational/whitened_variational_strategy.py]: Not sure I fully understand the `left_tensor` arg of inv_matmul - is this just for efficiency reasons in the context of CG/preconditioning?

File: test/examples/test_svgp_gp_regression.py [hunk 2]
```diff
@@ -74,26 +75,27 @@ def test_regression_error(self, cuda=False, skip_logdet_forward=False, cholesky=
...
-            self.assertTrue(param.grad is not None)
-            self.assertGreater(param.grad.norm().item(), 0)
-
-        # Set back to eval mode
-        model.eval()
-        likelihood.eval()
-        test_preds = likelihood(model(train_x)).mean.squeeze()
-        mean_abs_error = torch.mean(torch.abs(train_y - test_preds) / 2)
-        self.assertLess(mean_abs_error.item(), 1e-1)
+            self.assertFalse(any(w.category == ExtraComputationWarning for w in ws))
```
Comment (Balandat) [inline test/examples/test_svgp_gp_regression.py]: let's do `issubclass(w.category, ExtraComputationWarning)` here

File: test/examples/test_whitened_svgp_gp_regression.py [hunk 3]
```diff
@@ -75,26 +76,27 @@ def test_regression_error(self, cuda=False, skip_logdet_forward=False, cholesky=
...
-            self.assertTrue(param.grad is not None)
-            self.assertGreater(param.grad.norm().item(), 0)
-
-        # Set back to eval mode
-        model.eval()
-        likelihood.eval()
-        test_preds = likelihood(model(train_x)).mean.squeeze()
-        mean_abs_error = torch.mean(torch.abs(train_y - test_preds) / 2)
-        self.assertLess(mean_abs_error.item(), 1e-1)
+            self.assertFalse(any(w.category == ExtraComputationWarning for w in ws))
```
Comment (Balandat) [inline test/examples/test_whitened_svgp_gp_regression.py]: `issubclass`

File: test/lazy/test_cached_cg_lazy_tensor.py [hunk 4]
```diff
@@ -95,9 +95,9 @@ def _test_inv_matmul(self, rhs, lhs=None, cholesky=False):
                         res = lazy_tensor.inv_matmul(rhs)
                         actual = evaluated.inverse().matmul(rhs_copy)
                     self.assertAllClose(res, actual, rtol=0.02, atol=1e-5)
-                    self.assertEqual(len(w), 0)
+                    self.assertFalse(any(w.category == ExtraComputationWarning for w in ws))
```
Comment (Balandat) [inline test/lazy/test_cached_cg_lazy_tensor.py]: `issubclass`

File: test/lazy/test_cached_cg_lazy_tensor.py [hunk 5]
```diff
@@ -112,11 +112,11 @@ def _test_inv_matmul(self, rhs, lhs=None, cholesky=False):
                     # Determine if we've called CG or not
                     # We shouldn't if we supplied a lhs
                     if lhs is None:
-                        self.assertEqual(len(w), 1)
+                        self.assertEqual(len([w for w in ws if w.category == ExtraComputationWarning]), 1)
```
Comment (Balandat) [inline test/lazy/test_cached_cg_lazy_tensor.py]: `issubclass`

File: test/lazy/test_cached_cg_lazy_tensor.py [hunk 6]
```diff
@@ -112,11 +112,11 @@ def _test_inv_matmul(self, rhs, lhs=None, cholesky=False):
...
                         if not cholesky and self.__class__.should_call_cg:
                             self.assertTrue(linear_cg_mock.called)
                     else:
-                        self.assertEqual(len(w), 0)
+                        self.assertFalse(any(w.category == ExtraComputationWarning for w in ws))
```
Comment (Balandat) [inline test/lazy/test_cached_cg_lazy_tensor.py]: `issubclass`

File: test/lazy/test_cached_cg_lazy_tensor.py [hunk 7]
```diff
@@ -156,9 +156,9 @@ def test_inv_quad_logdet(self):
         vecs = lazy_tensor.eager_rhss[0].clone().detach().requires_grad_(True)
         vecs_copy = lazy_tensor.eager_rhss[0].clone().detach().requires_grad_(True)
 
-        with gpytorch.settings.num_trace_samples(128), warnings.catch_warnings(record=True) as w:
+        with gpytorch.settings.num_trace_samples(128), warnings.catch_warnings(record=True) as ws:  # noqa: F841
```
Comment (Balandat) [inline test/lazy/test_cached_cg_lazy_tensor.py]: remove the noqa

File: test/lazy/test_cached_cg_lazy_tensor.py [hunk 8]
```diff
@@ -156,9 +156,9 @@ def test_inv_quad_logdet(self):
...
             res_inv_quad, res_logdet = lazy_tensor.inv_quad_logdet(inv_quad_rhs=vecs, logdet=True)
-            self.assertEqual(len(w), 0)
+            self.assertFalse(any(w.category == ExtraComputationWarning for w in ws))
```
Comment (Balandat) [inline test/lazy/test_cached_cg_lazy_tensor.py]: `issubclass`

File: test/lazy/test_cached_cg_lazy_tensor.py [hunk 9]
```diff
@@ -179,11 +179,11 @@ def test_inv_quad_logdet_no_reduce(self):
         vecs = lazy_tensor.eager_rhss[0].clone().detach().requires_grad_(True)
         vecs_copy = lazy_tensor.eager_rhss[0].clone().detach().requires_grad_(True)
 
-        with gpytorch.settings.num_trace_samples(128), warnings.catch_warnings(record=True) as w:
+        with gpytorch.settings.num_trace_samples(128), warnings.catch_warnings(record=True) as ws:  # noqa: F841
```
Comment (Balandat) [inline test/lazy/test_cached_cg_lazy_tensor.py]: remove noqa

File: test/lazy/test_cached_cg_lazy_tensor.py [hunk 10]
```diff
@@ -179,11 +179,11 @@ def test_inv_quad_logdet_no_reduce(self):
...
             res_inv_quad, res_logdet = lazy_tensor.inv_quad_logdet(
                 inv_quad_rhs=vecs, logdet=True, reduce_inv_quad=False
             )
-            self.assertEqual(len(w), 0)
+            self.assertFalse(any(w.category == ExtraComputationWarning for w in ws))
```
Comment (Balandat) [inline test/lazy/test_cached_cg_lazy_tensor.py]: `issubclass`

**Pull request merged**